src/components/views/auth/CountryDropdown.tsx

```diff
@@ -0,0 +1,174 @@
+    const defaultCountry = defaultCountryCode ? COUNTRIES_BY_ISO2[defaultCountryCode] : COUNTRIES[0];
+    const [searchQuery, setSearchQuery] = React.useState('');
+    const { onOptionChange, value: baseValue } = props;
+
+    React.useEffect(() => {
+        if (baseValue) {
+            // If no value is given, we start with the default
+            // country selected, but our parent component
+            // doesn't know this, therefore we do this.
+            onOptionChange(defaultCountry);
+        }
+    }, [onOptionChange, defaultCountry, baseValue]);
+
+    let [displayedCountries, setDisplayedCountries] = React.useState<Country[]>([]);
+
+    if (searchQuery) {
+        // If we have a search query, we'll filter through the list of countries and render the result.
+        setDisplayedCountries(COUNTRIES.filter(
+            (country: Country) => countryMatchesSearchQuery(searchQuery, country),
+        ));
+
+        if (
+            searchQuery.length == 2 &&
+            COUNTRIES_BY_ISO2[searchQuery.toUpperCase()]
+        ) {
+            // The user is searching for an ISO2 code. Make the first result one
+            // that matches the ISO2.
+            const matched = COUNTRIES_BY_ISO2[searchQuery.toUpperCase()];
+
+            // Insert the matched item into the start of the array.
+            setDisplayedCountries([matched, ...displayedCountries.filter(country => country.iso2 != matched.iso2)]);
+        }
+    } else {
+        // Otherwise, if there isn't a search query, we'll just render every country in the list.
+        displayedCountries = COUNTRIES;
+    }
+
+    // Default value here too, otherwise we need to handle null / undefined
+    // values between mounting and the initial value propagating.
+    const value = props.value || defaultCountry.iso2;
+
+    const options = displayedCountries.map((country) => (
+        // Now it's time to take the options and render them.
+        <CountryDropdownOption {...props} value={value} country={country} key={country.iso2} />
+    ));
+
+    return (
+        <Dropdown
+            id="mx_CountryDropdown"
+            className={props.className + " mx_CountryDropdown"}
+            onOptionChange={(iso2) => onOptionChange(COUNTRIES_BY_ISO2[iso2])}
+            onSearchChange={(search) => setSearchQuery(search)}
+            menuWidth={298}
+            getShortOption={iso2 => <ShortOption {...props} value={iso2} />}
+            value={value}
+            searchEnabled={true}
+            disabled={props.disabled}
+            label={_t("Country Dropdown")}
+        >
+            { options }
+        </Dropdown>
+    );
+}
+
+CountryDropdown.propTypes = {
+    className: PropTypes.string,
+    isSmall: PropTypes.bool,
+    // if isSmall, show +44 in the selected value
+    showPrefix: PropTypes.bool,
+    onOptionChange: PropTypes.func.isRequired,
+    value: PropTypes.string,
+    disabled: PropTypes.bool,
+};
```

Comment:
we replace propTypes with Typescript interfaces

Reply:
See 3d83a8b and d2a02e4.

Reply:
This isn't resolved.

Reply:
Sorry! See https://github.com/matrix-org/matrix-react-sdk/pull/5285/commits/bb99f8d18cd6849434538f8a36e74f463caa6fd7